sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -42,15 +42,55 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     InternalRow.fromSeq(values.map(CatalystTypeConverters.convertToCatalyst))
   }
 
+  protected def convertToCatalystUnsafe(a: Any): Any = a match {
+    case arr: Array[Boolean] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Byte] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Short] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Int] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Long] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Float] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Double] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case other => CatalystTypeConverters.convertToCatalyst(other)
+  }
+
   protected def checkEvaluation(
       expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow): Unit = {
     val serializer = new JavaSerializer(new SparkConf()).newInstance
     val expr: Expression = serializer.deserialize(serializer.serialize(expression))
-    val catalystValue = CatalystTypeConverters.convertToCatalyst(expected)
+    // No codegen version expects GenericArrayData
+    val catalystValue = expected match {
+      case arr: Array[Byte] if expression.dataType == BinaryType => arr
+      case arr: Array[_] => new GenericArrayData(arr.map(CatalystTypeConverters.convertToCatalyst))
```

Comment:
In `CatalystTypeConverters.convertToCatalyst`, we have `case arr: Array[Any] => new GenericArrayData(arr.map(convertToCatalyst))`.

Actually it looks weird to me. Because most Array will not match this pattern and `convertToCatalyst` will return the original Array as it is. Looks it is not correct.

Can you try to change that pattern to `Array[_]`, and remove this line here? And we can see if it passes all tests.